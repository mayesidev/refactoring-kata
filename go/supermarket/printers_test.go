package supermarket

import (
	approvals "github.com/approvals/go-approval-tests"
	"github.com/approvals/go-approval-tests/reporters"
	"os"
	"testing"
)

func TestMain(m *testing.M) {
	r := approvals.UseReporter(reporters.NewContinuousIntegrationReporter())
	defer r.Close()

	os.Exit(m.Run())
}

type LineItem struct {
	product    Product
	quantity   float64
	unitPrice  float64
	totalPrice float64
}

func TestReceiptPrinter(t *testing.T) {
	var toothbrush = Product{name: "toothbrush", unit: Each}
	var apples = Product{name: "apples", unit: Kilo}
	var printer = NewReceiptPrinter()

	t.Parallel()
	tests := []struct {
		name      string
		products  []LineItem
		discounts []Discount
	}{
		{
			name:      "one line item",
			products:  []LineItem{{product: toothbrush, quantity: 1, unitPrice: 0.99, totalPrice: 0.99}},
			discounts: []Discount{},
		},
		{
			name:      "two line items",
			products:  []LineItem{{product: toothbrush, quantity: 2, unitPrice: 0.99, totalPrice: 0.99 * 2}},
			discounts: []Discount{},
		},
		{
			name:      "looseWeight",
			products:  []LineItem{{product: apples, quantity: 2.3, unitPrice: 1.99, totalPrice: 1.99 * 2.3}},
			discounts: []Discount{},
		},
		{
			name: "total",
			products: []LineItem{{product: toothbrush, quantity: 1, unitPrice: 0.99, totalPrice: 0.99},
				{product: apples, quantity: 0.75, unitPrice: 1.99, totalPrice: 1.99 * 0.75}},
			discounts: []Discount{},
		},
		{
			name:     "discounts",
			products: []LineItem{{product: apples, quantity: 2.3, unitPrice: 1.99, totalPrice: 1.99 * 2.3}},
			discounts: []Discount{{
				product:        apples,
				description:    "3 for 2",
				discountAmount: -0.99,
			}},
		},
		{
			name: "wholeReceipt",
			products: []LineItem{{product: toothbrush, quantity: 1, unitPrice: 0.99, totalPrice: 0.99},
				{product: toothbrush, quantity: 2, unitPrice: 0.99, totalPrice: 0.99 * 2},
				{product: apples, quantity: 0.75, unitPrice: 1.99, totalPrice: 1.99 * 0.75},
			},
			discounts: []Discount{{
				product:        toothbrush,
				description:    "3 for 2",
				discountAmount: -0.99,
			}},
		},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			var receipt = NewReceipt()
			for _, pq := range tc.products {
				receipt.addProduct(pq.product, pq.quantity, pq.unitPrice, pq.totalPrice)

			}
			for _, d := range tc.discounts {
				receipt.addDiscount(d)
			}
			approvals.VerifyString(t, printer.printReceipt(receipt))
		})
	}

}
