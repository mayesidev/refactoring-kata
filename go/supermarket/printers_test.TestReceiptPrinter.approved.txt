
========================================

toothbrush                          0.99

Total:                              0.99

========================================

toothbrush                          1.98
  0.99 * 2

Total:                              1.98

========================================

apples                              4.58
  1.99 * 2.300

Total:                              4.58

========================================

apples                              1.49
  1.99 * 0.750
toothbrush                          0.99

Total:                              2.48

========================================

apples                              4.58
  1.99 * 2.300
3 for 2 (apples)                   -0.99

Total:                              3.59

========================================

apples                              1.49
  1.99 * 0.750
toothbrush                          0.99
toothbrush                          1.98
  0.99 * 2
3 for 2 (toothbrush)               -0.99

Total:                              3.47
